packages/shorebird_cli/test/src/auth/auth_test.dart

```diff
@@ -38,11 +42,79 @@ void main() {
       auth = Auth(
         httpClient: httpClient,
         obtainAccessCredentials: (clientId, scopes, client, userPrompt) async {
-          return credentials;
+          return validCredentials;
         },
       )..logout();
     });
 
+    group('AuthenticatedClient', () {
+      test('refreshes and uses new token when credentials are expired.',
+          () async {
+        when(() => httpClient.send(any())).thenAnswer(
+          (_) async => http.StreamedResponse(
+            const Stream.empty(),
+            HttpStatus.ok,
+          ),
+        );
+
+        final refreshCallbacks = <AccessCredentials>[];
+        final expiredCredentials = AccessCredentials(
+          AccessToken(
+            'Bearer',
+            'accessToken',
+            DateTime.now().subtract(const Duration(minutes: 1)).toUtc(),
+          ),
+          '',
+          [],
+          idToken: 'expiredIdToken',
+        );
+
+        final client = AuthenticatedClient(
+          credentials: expiredCredentials,
+          httpClient: httpClient,
+          onRefreshCallback: refreshCallbacks.add,
+          refreshCredentials: (clientId, credentials, client) async =>
+              validCredentials,
+        );
+
+        await client.get(Uri.parse('https://example.com'));
+
+        expect(
+          refreshCallbacks,
+          equals([
+            isA<AccessCredentials>().having((c) => c.idToken, 'token', idToken)
+          ]),
+        );
+        final captured = verify(() => httpClient.send(captureAny())).captured;
+        expect(captured, hasLength(1));
+        final request = captured.first as http.BaseRequest;
+        expect(request.headers['Authorization'], equals('Bearer $idToken'));
+      });
+
+      test('uses valid token when credentials valid.', () async {
+        when(() => httpClient.send(any())).thenAnswer(
+          (_) async => http.StreamedResponse(
+            const Stream.empty(),
+            HttpStatus.ok,
+          ),
+        );
+        final refreshCallbacks = <AccessCredentials>[];
+        final client = AuthenticatedClient(
+          credentials: validCredentials,
+          httpClient: httpClient,
+          onRefreshCallback: refreshCallbacks.add,
+        );
+
+        await client.get(Uri.parse('https://example.com'));
+
+        expect(refreshCallbacks, isEmpty);
+        final captured = verify(() => httpClient.send(captureAny())).captured;
+        expect(captured, hasLength(1));
+        final request = captured.first as http.BaseRequest;
+        expect(request.headers['Authorization'], equals('Bearer $idToken'));
```

Comment:
Should we also verify that the refresh callback was never invoked?

Reply:
I do that on line 110

Reply:
Missed it! Ignore me :)

Reply:
No worries glad you asked!